terraform/eval_variable.go

```diff
@@ -2,12 +2,58 @@ package terraform
 
 import (
 	"fmt"
+	"strings"
 
 	"github.com/hashicorp/errwrap"
 	"github.com/hashicorp/terraform/config"
+	"github.com/hashicorp/terraform/config/module"
 	"github.com/mitchellh/mapstructure"
 )
 
+type EvalTypeCheckVariable struct {
```

Comment:
Yes that seems like a worthwhile effort - I think several others can be documented based on the research that went into this PR too!